### PR TITLE
Fix frame pointer is clobbered before saving vet issue on BP register

### PR DESCRIPTION
**Pull request by zchee**
Fix frame pointer is clobbered before saving vet issue on `BP` register.

`BP` should be callee-save. It will be saved automatically if there is nonzero frame size.

```console
$ go1.16.5 vet -framepointer ./pkg/atomicbitops
# gvisor.dev/gvisor/pkg/atomicbitops
atomicbitops/atomicbitops_amd64.s:20:1: frame pointer is clobbered before saving
atomicbitops/atomicbitops_amd64.s:27:1: frame pointer is clobbered before saving
atomicbitops/atomicbitops_amd64.s:34:1: frame pointer is clobbered before saving
atomicbitops/atomicbitops_amd64.s:50:1: frame pointer is clobbered before saving
atomicbitops/atomicbitops_amd64.s:57:1: frame pointer is clobbered before saving
atomicbitops/atomicbitops_amd64.s:64:1: frame pointer is clobbered before saving
```

### CODE REVIEW

**Comment by avagin** [comment]
@prattmic could you take a look at this pr?

**Comment by zchee** [comment]
another way is push and pop BP

File: pkg/atomicbitops/atomicbitops_amd64.s [hunk 1]
```diff
@@ -16,21 +16,21 @@
 
 #include "textflag.h"
 
-TEXT ·AndUint32(SB),$0-12
+TEXT ·AndUint32(SB),$8-12
   MOVQ  addr+0(FP), BP
```
Comment (prattmic) [inline pkg/atomicbitops/atomicbitops_amd64.s:20]: Why do these functions use BP at all? I'm pretty sure this is just a typo for BX, and we should just change to BX.

**Review by prattmic — CHANGES_REQUESTED** [review]

File: pkg/atomicbitops/atomicbitops_amd64.s (hunk 1 above)
Comment (zchee) [reply pkg/atomicbitops/atomicbitops_amd64.s:20]: I see. I'll fix BP to BX at all.

**Comment by zchee** [comment]
@prattmic Fixed to your suggestion, but test failed:

```
=== RUN   TestOrUint32
    atomicbitops_test.go:68: Data race detected!
--- FAIL: TestOrUint32 (0.00s)
=== RUN   TestAndUint32
    atomicbitops_test.go:74: Data race detected!
--- FAIL: TestAndUint32 (0.00s)
=== RUN   TestXorUint32
    atomicbitops_test.go:80: Data race detected!
--- FAIL: TestXorUint32 (0.00s)
=== RUN   TestOrUint64
    atomicbitops_test.go:86: Data race detected!
--- FAIL: TestOrUint64 (0.00s)
=== RUN   TestAndUint64
    atomicbitops_test.go:92: Data race detected!
--- FAIL: TestAndUint64 (0.00s)
=== RUN   TestXorUint64
    atomicbitops_test.go:98: Data race detected!
--- FAIL: TestXorUint64 (0.00s)
=== RUN   TestCompareAndSwapUint32
--- PASS: TestCompareAndSwapUint32 (0.00s)
=== RUN   TestCompareAndSwapUint64
--- PASS: TestCompareAndSwapUint64 (0.00s)
FAIL
FAIL	gvisor.dev/gvisor/pkg/atomicbitops	0.006s
FAIL
```

File: pkg/atomicbitops/atomicbitops_amd64.s [hunk 2]
```diff
@@ -17,21 +17,21 @@
 #include "textflag.h"
 
 TEXT ·AndUint32(SB),$0-12
-  MOVQ  addr+0(FP), BP
+  MOVQ  addr+0(FP), BX
   MOVL  val+8(FP), AX
   LOCK
   ANDL   AX, 0(BP)
```
Comment (prattmic) [inline pkg/atomicbitops/atomicbitops_amd64.s:23]: You need to update these as well.

Comment (zchee) [reply pkg/atomicbitops/atomicbitops_amd64.s:23]: ah, forgot. FIx ASAP.

**Review by prattmic — APPROVED** [review]
LGTM

**Comment by zchee** [comment]
@prattmic Hi,
still copybara not pulled to upstream, Is this change still not approved by Google internal?

**Comment by prattmic** [comment]
@zchee Sorry for the delay! This change got missed.

**Pull request merged**